### DISCUSSION TITLE
Wrong current balance

### DISCUSSION THREAD

**Discussion by Jaeger87** [firefly-iii/firefly-iii#8919] [Firefly III]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

Hi,

One of my asset accounts has an incorrect balance value (0.01 instead of 0). When I go into reconciliation mode for that account, if I leave the balance at 0.01, Firefly suggests a correction. However, if I set the balance to zero, Firefly tells me there is nothing to do.

This issue occurred after a problem with a recurring transaction. After the transactions were executed, Firefly became unstable and gave me an error 500 when I opened the recurring transaction page. After rebooting the Docker container, everything seemed fine except for the incorrect balance in that account.

Do you have any suggestions?

### Debug information

Debug information generated at 2024-05-24 17:50:15 for Firefly III version **v6.1.16**.

<table><thead><tr><th colspan="2">System information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>Firefly III</td><td>6.1.16 / v2.1.0 / 24 (exp. 24)</td></tr><tr><td>PHP version</td><td>8.3.6 (64bits) / apache2handler / Linux x86_64</td></tr><tr><td>BCscale</td><td>12</td></tr><tr><td>Error reporting</td><td>Display: Off, reporting: ALL errors</td></tr><tr><td>Max upload</td><td>67108864 (64 MB)</td></tr><tr><td>Database drivers</td><td><strong>*mysql*</strong>, pgsql, sqlite, </td></tr><tr><td>Docker build</td><td><span>#</span>1086, base <span>#</span>83</td></tr></tbody></table><table><thead><tr><th colspan="2">Firefly III information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>Timezone</td><td>UTC + Europe/Rome</td></tr><tr><td>App environment</td><td>production, debug: false</td></tr><tr><td>Layout</td><td>v1</td></tr><tr><td>Logging</td><td>info, stack / (empty)</td></tr><tr><td>Cache driver</td><td>file</td></tr><tr><td>Default language and locale</td><td>en_US + equal</td></tr><tr><td>Trusted proxies</td><td>**</td></tr><tr><td>Login provider & user guard</td><td>eloquent / web</td></tr><tr><td>Login headers</td><td>N/A + N/A</td></tr><tr><td>Stateful domains</td><td></td></tr><tr><td>Last cron job</td><td>2024-05-24 03:00:00 (14 hours ago)</td></tr><tr><td>Mailer</td><td>log</td></tr></tbody></table><table><thead><tr><th colspan="2">User-specific information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>User</td><td><span>#</span>1 of 1</td></tr><tr><td>User flags</td><td><span title="Has liabilities">:credit_card:</span> <span title="Has rules">:wrench:</span> <span title="Has recurring transactions">:clock130:</span></td></tr><tr><td>Session start</td><td>2024-05-01 00:00:00</td></tr><tr><td>Session end</td><td>2024-05-31 23:59:59</td></tr><tr><td>View range</td><td>1M</td></tr><tr><td>User language</td><td>it_IT</td></tr><tr><td>User locale</td><td>it_IT</td></tr><tr><td>Locale(s) supported</td><td>it_IT.utf8: :white_check_mark:<br>it_IT.UTF-8: :white_check_mark:<br></td></tr><tr><td>User agent</td><td>Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:126.0) Gecko/20100101 Firefox/126.0</td></tr></tbody></table>

**Comment by JC5** [comment]
Make a backup of your database, and then run `docker exec -it [yourcontainer] php artisan firefly-iii:force-decimal-size`. That should fix it.

**Comment by Jaeger87** [comment]
Ok, thank you. I'll try tomorrow!

Regarding the automation problem, even this morning Firefly entered a bad status, and I had to restart the container to restore Firefly III. This is the command I put in the cron job: '/usr/local/bin/php /var/www/html/artisan firefly-iii:cron', and it is executed by the [Cosmos ](https://github.com/azukaar/Cosmos-Server)scheduler.

This is the error message (sorry, it's in Italian):

```
Sfortunatamente questo errore non è riparabile :(. Firefly III è rotto. L'errore è:

Allowed memory size of 1073741824 bytes exhausted (tried to allocate 375407696 bytes)

Questo errore si è verificato nel file /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Exceptions/Handler.php alla riga 802 con codice 0.
```
And this is the log from the cron job:

```
[2024-05-25 03:00:00] production.INFO: It has been 21h after since the recurring transactions cron-job has fired. It will fire now!  
o[2024-05-25 03:00:00] production.INFO: Will now fire recurring cron job task for date "2024-05-25 03:00:00".  
}[2024-05-25 03:00:00] production.INFO: Recurrence #2 is set to run on 2024-06-01, and today's date is 2024-05-25. Skipped.  
}[2024-05-25 03:00:00] production.INFO: Recurrence #3 is set to run on 2024-06-01, and today's date is 2024-05-25. Skipped.  
}[2024-05-25 03:00:00] production.INFO: Recurrence #1 is set to run on 2024-06-01, and today's date is 2024-05-25. Skipped.  
}[2024-05-25 03:00:00] production.INFO: Recurrence #4 is set to run on 2024-05-26, and today's date is 2024-05-25. Skipped.  
~[2024-05-25 03:00:00] production.INFO: Recurrence #14 is set to run on 2024-07-24, and today's date is 2024-05-25. Skipped.  
~[2024-05-25 03:00:00] production.INFO: Recurrence #11 is set to run on 2024-06-02, and today's date is 2024-05-25. Skipped.  
~[2024-05-25 03:00:00] production.INFO: Recurrence #13 is set to run on 2024-06-24, and today's date is 2024-05-25. Skipped.  
u[2024-05-25 03:00:00] production.INFO: Marked the last time this job has run as "2024-05-25 03:00:00" (1716606000)  
L[2024-05-25 03:00:00] production.INFO: Done with recurring cron job task.  
  [i] "Create recurring transactions" cron fired: Recurring transactions cron job fired successfully.
  [✓] "Create recurring transactions" cron ran with success: Recurring transactions cron job fired successfully.
{[2024-05-25 03:00:00] production.INFO: It has been 21h after since the auto budget cron-job has fired. It will fire now!  
h[2024-05-25 03:00:00] production.INFO: Will now fire auto budget cron job task for date "2024-05-25".  
N[2024-05-25 03:00:00] production.INFO: Done with auto budget cron job task.  
  [i] "Create auto budgets" cron fired: Auto-budget cron job fired successfully.
  [✓] "Create auto budgets" cron ran with success: Auto-budget cron job fired successfully.
|[2024-05-25 03:00:00] production.INFO: It has been 21h after since the bill warning cron-job has fired. It will fire now!  
m[2024-05-25 03:00:00] production.INFO: Will now fire bill warning job task for date "2024-05-25 03:00:00".  
[2024-05-25 03:00:00] production.INFO: Marked the last time this job has run as "2024-05-25 03:00:00" (1716606000)  
[2024-05-25 03:00:00] production.INFO: Done with bill warning cron job task.  
Q  [i] "Send bill warnings" cron fired: Bill warning cron job fired successfully.
^  [✓] "Send bill warnings" cron ran with success: Bill warning cron job fired successfully.
D  [i] More feedback on the cron jobs can be found in the log files.
```

If you need other information let me know it. Thank you!

**Comment by Jaeger87** [comment]
I tried to use your command, it ended with an error, this is the output:

`
[2024-05-27 19:30:42] production.ERROR: Exception is: {"class":"TypeError","errorMessage":"Unsupported operand types: string * int","time":"Mon, 27 May 2024 19:30:42 +0000","file":"\/var\/www\/html\/app\/Console\/Commands\/System\/ForceDecimalSize.php","line":315,"code":0,"version":"6.1.16","url":"https:\/\/192.168.1.58\/cosmos-Firefly-III-default","userAgent":"Symfony","json":true,"method":"GET","post":""}
[2024-05-27 19:30:42] production.ERROR: Unsupported operand types: string * int {"exception":"[object] (TypeError(code: 0): Unsupported operand types: string * int at /var/www/html/app/Console/Commands/System/ForceDecimalSize.php:315)
[stacktrace]
#0 /var/www/html/app/Console/Commands/System/ForceDecimalSize.php(194): FireflyIII\\Console\\Commands\\System\\ForceDecimalSize->correctGeneric()
#1 /var/www/html/app/Console/Commands/System/ForceDecimalSize.php(161): FireflyIII\\Console\\Commands\\System\\ForceDecimalSize->correctByCurrency()
#2 /var/www/html/app/Console/Commands/System/ForceDecimalSize.php(146): FireflyIII\\Console\\Commands\\System\\ForceDecimalSize->correctAmountsByCurrency()
#3 /var/www/html/app/Console/Commands/System/ForceDecimalSize.php(103): FireflyIII\\Console\\Commands\\System\\ForceDecimalSize->correctAmounts()
#4 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): FireflyIII\\Console\\Commands\\System\\ForceDecimalSize->handle()
#5 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
#6 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\\Container\\Util::unwrapIfClosure()
#7 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(35): Illuminate\\Container\\BoundMethod::callBoundMethod()
#8 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(662): Illuminate\\Container\\BoundMethod::call()
#9 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(212): Illuminate\\Container\\Container->call()
#10 /var/www/html/vendor/symfony/console/Command/Command.php(279): Illuminate\\Console\\Command->execute()
#11 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(181): Symfony\\Component\\Console\\Command\\Command->run()
#12 /var/www/html/vendor/symfony/console/Application.php(1049): Illuminate\\Console\\Command->run()
#13 /var/www/html/vendor/symfony/console/Application.php(318): Symfony\\Component\\Console\\Application->doRunCommand()
#14 /var/www/html/vendor/symfony/console/Application.php(169): Symfony\\Component\\Console\\Application->doRun()
#15 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(196): Symfony\\Component\\Console\\Application->run()
#16 /var/www/html/artisan(35): Illuminate\\Foundation\\Console\\Kernel->handle()
#17 {main}
"}

   TypeError

  Unsupported operand types: string * int

  at app/Console/Commands/System/ForceDecimalSize.php:315
    311▕                     continue;
    312▕                 }
    313▕                 // fix $field by rounding it down correctly.
    314▕                 $pow     = 10 ** $currency->decimal_places;
  ➜ 315▕                 $correct = bcdiv((string)round($value * $pow), (string)$pow, 12);
    316▕                 $this->friendlyWarning(sprintf('%s #%d has %s with value "%s", this has been corrected to "%s".', $table, $item->id, $field, $value, $correct));
    317▕                 $class::find($item->id)->update([$field => $correct]);
    318▕             }
    319▕         }

  1   app/Console/Commands/System/ForceDecimalSize.php:194
      FireflyIII\Console\Commands\System\ForceDecimalSize::correctGeneric()

  2   app/Console/Commands/System/ForceDecimalSize.php:161
      FireflyIII\Console\Commands\System\ForceDecimalSize::correctByCurrency()
`

**Reply by JC5** [reply]
Mmm, that's an indication you have more serious database issues. I can catch that error but it may mean your problem will not be fixed easily. The next development release will have an improved command for this, but I'm not sure if Cosmos will allow you to install it.

**Comment by Jaeger87** [comment]
Is it possible that one or more transactions are causing this problem? This is a brand new Firefly III installation (I installed Firefly this month and it contains my money movements for 2024), so if I have a clue about the problem, I can search on my own to find which transaction is creating this issue.

Regarding the memory problem, every morning I have to restart Firefly III myself (I'm thinking about creating a cron job to do this). I don't know if this problem is linked to the incorrect current balance; I suspect it could be the source, but it's just a suspicion.

**Comment by Jaeger87** [comment]
I don't know if it's a clue, but I opened the MariaDB logs and found this:

2024-05-23 07:38:51
2024-05-23 7:38:51 8338 [Warning] Aborted connection 8338 to db: 'unconnected' user: 'unauthenticated' host: '172.16.0.11' (This connection closed normally without authentication)
2024-05-24 07:00:52
2024-05-24 7:00:52 10166 [Warning] Aborted connection 10166 to db: 'unconnected' user: 'unauthenticated' host: '172.16.0.11' (This connection closed normally without authentication)
2024-05-25 12:15:15
2024-05-25 12:15:15 12566 [Warning] Aborted connection 12566 to db: 'unconnected' user: 'unauthenticated' host: '172.16.0.11' (This connection closed normally without authentication)
2024-05-26 15:56:00
2024-05-26 15:56:00 14355 [Warning] Aborted connection 14355 to db: 'unconnected' user: 'unauthenticated' host: '172.16.0.11' (This connection closed normally without authentication)

Could this be related to the automation cron job?

Just for a test, I turned off the cron job for the automations, so if Firefly III doesn't get stuck tomorrow, it's definitely an issue with the automations.

**Reply by JC5** [reply]
> Could this be related to the automation cron job?

No, that's from the startup routine of Firefly III, pinging the DB to see if it is up and running.

**Reply by Jaeger87** [reply]
Ok, probably that line gets written every time I restart Firefly III.

I made the test with the cron job turned off,  and as expected, there were no issues with Firefly III.
